### ISSUE TITLE
understanding why links aren't getting marked as visited...

### ISSUE THREAD

**Issue by lgarcia-tritier** [MicrosoftEdge/WebView2Feedback#2378]
Windows Forms MDI application, and currently working in moving from the WebBrowser control onto WebView2.
I've basically unplugged WebBrowser and plugged WebView2 (wrapped it), plus making sure all parts are properly initialized, during this update.

The applications uses HTML web pages to allow searches and display results.
How it works: The initial search screen is created locally and displayed via the control. The application listens for events, parses the request and sends it to a set of Windows Services on a server which gather the results and send back html to the client. The client then 'navigates' to this html that is received...rinse and repeat. All HTML pages are saved to disk prior to calling navigate.

Without any specific effort, WebBrowser was keeping track of visited links which would automatically get marked and this history was kept across sessions and applications restarts. With WebView2, I have been unable to determine why visited links aren't being styled as such, not even during the current session.

I tried rolling out my own 'visited' style class, but that doesn't meet the requirement since visited links are only marked as such during the current session. I also tried using a virtual host (https://github.com/MicrosoftEdge/WebView2Feedback/issues/1891), but got no better results. I've tried to override the style, but nay...it is as if something was overriding the style, I don't see that in the element inspector...I expected this to work out of the box, I must be missing or not understanding something around history.

Any pointers?

**Comment by yildirimcagri** [comment]
Hi @lgarcia-tritier, I do see visited links highlighted in purple when I browse Bing in WebView2, persisting through different sessions as well. However, in this case, sites could be keeping visited link history and style the links as "visited" themselves. Since WebView2 is based on Microsoft Edge, the behavior should match Edge here, is that what you see as well with your web pages?

**Comment by lgarcia-tritier** [comment]
Thanks for the answer. Yes, I went back to the basics by using one of the WebView2 samples, and indeed I saw the links get marked as visited as one would expect.

I went back to using the WebBrowser control to make sure I hadn't broken the workflow and links did get marked as visited.
Clearly, there's a difference in how the WebBrowser control was handling, but it may also be a case of how the pages are getting served in this application which worked with IE, but not fully with Edge/Chrome. I continue to look into this...

Once I have something more interesting, or a workaround, I will post again.

**Comment by lgarcia-tritier** [comment]
I spent some time fighting with this issue and I think the reason might be the way this MDI application uses the web control to navigate to URLs, and then there sure must be a difference on how WebView2 track of these vs WebBrowser.

A page with a link isn't directly tied to the next navigation page, meaning the URLs after parsing and processing, are ultimately different. 

For example: http://Link_A is clicked, the URL is then parsed by the application and processed, then the browser receives a new URL file://page to navigate to. I suspect this gap is the reason why it isn't tracked as expected.

Re-architecting to test this theory isn't an option at this time, so I opted to roll-out my own style for visited links and then keeping track of clicked on links in memory as a workaround. I suppose I could have tried to look into updating the history, but this seems to work good enough for now.
